### PR TITLE
added dino

### PR DESCRIPTION
**Pull request by DonkeyShot21**

### CODE REVIEW

File: solo/losses/dino.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+        total_loss /= n_loss_terms
+        self.update_center(teacher_output)
+        return total_loss
+
+    @torch.no_grad()
+    def update_center(self, teacher_output):
+        batch_center = torch.sum(teacher_output, dim=0, keepdim=True)
+        if dist.is_available() and dist.is_initialized():
+            dist.all_reduce(batch_center)
+            batch_center = batch_center / (len(teacher_output) * dist.get_world_size())
```
Comment (vturrisi) [inline solo/losses/dino.py:63]: should this be inside the if or outside?

File: solo/methods/dino.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+            self.mlp = nn.Sequential(*layers)
+        self.apply(self._init_weights)
+
+        self.last_layer = nn.utils.weight_norm(nn.Linear(bottleneck_dim, out_dim, bias=False))
+        self.last_layer.weight_g.data.fill_(1)
+
+        if norm_last_layer:
+            self.last_layer.weight_g.requires_grad = False
+
+    def _init_weights(self, m):
```
Comment (vturrisi) [inline solo/methods/dino.py:46]: does this changes how the weights of the resnet are initialised?

File: solo/methods/dino.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+        if isinstance(m, nn.Linear):
+            trunc_normal_(m.weight, std=0.02)
+            if isinstance(m, nn.Linear) and m.bias is not None:
+                nn.init.constant_(m.bias, 0)
+
+    def forward(self, x):
+        x = self.mlp(x)
+        x = nn.functional.normalize(x, dim=-1, p=2)
```
Comment (vturrisi) [inline solo/methods/dino.py:54]: change to x = F.normalize(x)

File: solo/methods/dino.py [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+
+    def clip_gradients(self, clip):
+        for p in self.encoder.parameters():
+            if p.grad is not None:
+                param_norm = p.grad.data.norm(2)
+                clip_coef = clip / (param_norm + 1e-6)
+                if clip_coef < 1:
+                    p.grad.data.mul_(clip_coef)
+
+    def on_train_epoch_start(self,):
```
Comment (vturrisi) [inline solo/methods/dino.py:149]: remove ,

**Review by vturrisi — APPROVED** [review]
